File: backend/app/api/applications.py
import uuid

from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.orm import Session

from app.database import get_db
from app.schemas.applications import (
    ApplicationDetailResponse,
    ApplicationExternalLinkCreate,
    ApplicationExternalLinkResponse,
    ApplicationExternalLinkUpdate,
    ApplicationInterviewCancelRequest,
    ApplicationInterviewCompleteRequest,
    ApplicationInterviewStageCreate,
    ApplicationInterviewStageResponse,
    ApplicationInterviewStageUpdate,
    ApplicationListItemResponse,
    ApplicationMetadataUpdate,
    ApplicationNoteCreate,
    ApplicationNoteResponse,
    ApplicationNoteUpdate,
    ApplicationPipelineResponse,
    ApplicationStateTransitionRequest,
    ApplicationTimelineEventResponse,
)
from app.services.applications import (
    ApplicationWorkflowNotFoundError,
    ApplicationWorkflowSeedMissingError,
    ApplicationWorkflowService,
    ApplicationWorkflowTransitionError,
    ApplicationWorkflowValidationError,
    ApplicationWorkflowWorkspaceInactiveError,
    ApplicationWorkflowWorkspaceNotFoundError,
)

router = APIRouter(tags=["applications"])


def get_application_workflow_service(
    db: Session = Depends(get_db),
) -> ApplicationWorkflowService:
    return ApplicationWorkflowService(db)


@router.post(
    "/roles/{role_id}/application",
    response_model=ApplicationDetailResponse,
    status_code=status.HTTP_201_CREATED,
)
def get_or_create_role_application(
    role_id: uuid.UUID,
    service: ApplicationWorkflowService = Depends(get_application_workflow_service),
):
    try:
        return service.get_or_create_for_role(role_id)
    except ApplicationWorkflowNotFoundError as exc:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(exc))
    except ApplicationWorkflowWorkspaceInactiveError as exc:
        raise HTTPException(status_code=status.HTTP_409_CONFLICT, detail=str(exc))
    except ApplicationWorkflowValidationError as exc:
        raise HTTPException(status_code=status.HTTP_502_BAD_GATEWAY, detail=str(exc))
    except ApplicationWorkflowSeedMissingError as exc:
        raise HTTPException(status_code=status.HTTP_409_CONFLICT, detail=str(exc))


@router.get("/applications", response_model=list[ApplicationListItemResponse])
def list_applications(
    workspace_id: uuid.UUID | None = None,
    include_inactive: bool = False,
    service: ApplicationWorkflowService = Depends(get_application_workflow_service),
):
    try:
        return service.list_applications(
            workspace_id=workspace_id,
            include_inactive=include_inactive,
        )
    except ApplicationWorkflowWorkspaceNotFoundError as exc:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(exc))
    except ApplicationWorkflowValidationError as exc:
        raise HTTPException(status_code=status.HTTP_502_BAD_GATEWAY, detail=str(exc))
    except ApplicationWorkflowSeedMissingError as exc:
        raise HTTPException(status_code=status.HTTP_409_CONFLICT, detail=str(exc))


@router.get(
    "/workspaces/{workspace_id}/applications",
    response_model=list[ApplicationListItemResponse],
)
def list_workspace_applications(
    workspace_id: uuid.UUID,
    include_inactive: bool = False,
    service: ApplicationWorkflowService = Depends(get_application_workflow_service),
):
    try:
        return service.list_applications(
            workspace_id=workspace_id,
            include_inactive=include_inactive,
        )
    except ApplicationWorkflowWorkspaceNotFoundError as exc:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(exc))
    except ApplicationWorkflowValidationError as exc:
        raise HTTPException(status_code=status.HTTP_502_BAD_GATEWAY, detail=str(exc))
    except ApplicationWorkflowSeedMissingError as exc:
        raise HTTPException(status_code=status.HTTP_409_CONFLICT, detail=str(exc))


@router.get("/applications/pipeline", response_model=ApplicationPipelineResponse)
def get_applications_pipeline(
    workspace_id: uuid.UUID | None = None,
    include_inactive: bool = False,
    service: ApplicationWorkflowService = Depends(get_application_workflow_service),
):
    try:
        return service.get_pipeline(
            workspace_id=workspace_id,
            include_inactive=include_inactive,
        )
    except ApplicationWorkflowWorkspaceNotFoundError as exc:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(exc))
    except ApplicationWorkflowValidationError as exc:
        raise HTTPException(status_code=status.HTTP_502_BAD_GATEWAY, detail=str(exc))
    except ApplicationWorkflowSeedMissingError as exc:
        raise HTTPException(status_code=status.HTTP_409_CONFLICT, detail=str(exc))


@router.get(
    "/workspaces/{workspace_id}/applications/pipeline",
    response_model=ApplicationPipelineResponse,
)
def get_workspace_applications_pipeline(
    workspace_id: uuid.UUID,
    include_inactive: bool = False,
    service: ApplicationWorkflowService = Depends(get_application_workflow_service),
):
    try:
        return service.get_pipeline(
            workspace_id=workspace_id,
            include_inactive=include_inactive,
        )
    except ApplicationWorkflowWorkspaceNotFoundError as exc:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(exc))
    except ApplicationWorkflowValidationError as exc:
        raise HTTPException(status_code=status.HTTP_502_BAD_GATEWAY, detail=str(exc))
    except ApplicationWorkflowSeedMissingError as exc:
        raise HTTPException(status_code=status.HTTP_409_CONFLICT, detail=str(exc))


@router.get("/applications/{application_id}", response_model=ApplicationDetailResponse)
def get_application(
    application_id: uuid.UUID,
    service: ApplicationWorkflowService = Depends(get_application_workflow_service),
):
    try:
        return service.get_application(application_id)
    except ApplicationWorkflowNotFoundError as exc:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(exc))
    except ApplicationWorkflowValidationError as exc:
        raise HTTPException(status_code=status.HTTP_502_BAD_GATEWAY, detail=str(exc))
    except ApplicationWorkflowSeedMissingError as exc:
        raise HTTPException(status_code=status.HTTP_409_CONFLICT, detail=str(exc))


@router.get(
    "/applications/{application_id}/timeline",
    response_model=list[ApplicationTimelineEventResponse],
)
def get_application_timeline(
    application_id: uuid.UUID,
    service: ApplicationWorkflowService = Depends(get_application_workflow_service),
):
    try:
        return service.get_timeline(application_id)
    except ApplicationWorkflowNotFoundError as exc:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(exc))
    except ApplicationWorkflowSeedMissingError as exc:
        raise HTTPException(status_code=status.HTTP_409_CONFLICT, detail=str(exc))


@router.get(
    "/applications/{application_id}/notes",
    response_model=list[ApplicationNoteResponse],
)
def list_application_notes(
    application_id: uuid.UUID,
    service: ApplicationWorkflowService = Depends(get_application_workflow_service),
):
    try:
        return service.list_notes(application_id)
    except ApplicationWorkflowNotFoundError as exc:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(exc))
    except ApplicationWorkflowSeedMissingError as exc:
        raise HTTPException(status_code=status.HTTP_409_CONFLICT, detail=str(exc))


@router.post(
    "/applications/{application_id}/notes",
    response_model=ApplicationNoteResponse,
    status_code=status.HTTP_201_CREATED,
)
def create_application_note(
    application_id: uuid.UUID,
    payload: ApplicationNoteCreate,
    service: ApplicationWorkflowService = Depends(get_application_workflow_service),
):
    try:
        return service.create_note(application_id, payload)
    except ApplicationWorkflowNotFoundError as exc:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(exc))
    except ApplicationWorkflowSeedMissingError as exc:
        raise HTTPException(status_code=status.HTTP_409_CONFLICT, detail=str(exc))


@router.patch(
    "/applications/{application_id}/notes/{note_id}",
    response_model=ApplicationNoteResponse,
)
def update_application_note(
    application_id: uuid.UUID,
    note_id: uuid.UUID,
    payload: ApplicationNoteUpdate,
    service: ApplicationWorkflowService = Depends(get_application_workflow_service),
):
    try:
        return service.update_note(application_id, note_id, payload)
    except ApplicationWorkflowNotFoundError as exc:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(exc))
    except ApplicationWorkflowSeedMissingError as exc:
        raise HTTPException(status_code=status.HTTP_409_CONFLICT, detail=str(exc))


@router.delete(
    "/applications/{application_id}/notes/{note_id}",
    status_code=status.HTTP_204_NO_CONTENT,
)
def delete_application_note(
    application_id: uuid.UUID,
    note_id: uuid.UUID,
    service: ApplicationWorkflowService = Depends(get_application_workflow_service),
):
    try:
        service.delete_note(application_id, note_id)
    except ApplicationWorkflowNotFoundError as exc:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(exc))
    except ApplicationWorkflowSeedMissingError as exc:
        raise HTTPException(status_code=status.HTTP_409_CONFLICT, detail=str(exc))


@router.get(
    "/applications/{application_id}/interviews",
    response_model=list[ApplicationInterviewStageResponse],
)
def list_application_interviews(
    application_id: uuid.UUID,
    service: ApplicationWorkflowService = Depends(get_application_workflow_service),
):
    try:
        return service.list_interview_stages(application_id)
    except ApplicationWorkflowNotFoundError as exc:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(exc))
    except ApplicationWorkflowSeedMissingError as exc:
        raise HTTPException(status_code=status.HTTP_409_CONFLICT, detail=str(exc))


@router.post(
    "/applications/{application_id}/interviews",
    response_model=ApplicationInterviewStageResponse,
    status_code=status.HTTP_201_CREATED,
)
def create_application_interview(
    application_id: uuid.UUID,
    payload: ApplicationInterviewStageCreate,
    service: ApplicationWorkflowService = Depends(get_application_workflow_service),
):
    try:
        return service.create_interview_stage(application_id, payload)
    except ApplicationWorkflowNotFoundError as exc:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(exc))
    except ApplicationWorkflowValidationError as exc:
        raise HTTPException(status_code=status.HTTP_422_UNPROCESSABLE_ENTITY, detail=str(exc))
    except ApplicationWorkflowSeedMissingError as exc:
        raise HTTPException(status_code=status.HTTP_409_CONFLICT, detail=str(exc))


@router.patch(
    "/applications/{application_id}/interviews/{interview_id}",
    response_model=ApplicationInterviewStageResponse,
)
def update_application_interview(
    application_id: uuid.UUID,
    interview_id: uuid.UUID,
    payload: ApplicationInterviewStageUpdate,
    service: ApplicationWorkflowService = Depends(get_application_workflow_service),
):
    try:
        return service.update_interview_stage(application_id, interview_id, payload)
    except ApplicationWorkflowNotFoundError as exc:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(exc))
    except ApplicationWorkflowValidationError as exc:
        raise HTTPException(status_code=status.HTTP_422_UNPROCESSABLE_ENTITY, detail=str(exc))
    except ApplicationWorkflowSeedMissingError as exc:
        raise HTTPException(status_code=status.HTTP_409_CONFLICT, detail=str(exc))


@router.post(
    "/applications/{application_id}/interviews/{interview_id}/complete",
    response_model=ApplicationInterviewStageResponse,
)
def complete_application_interview(
    application_id: uuid.UUID,
    interview_id: uuid.UUID,
    payload: ApplicationInterviewCompleteRequest | None = None,
    service: ApplicationWorkflowService = Depends(get_application_workflow_service),
):
    try:
        return service.complete_interview_stage(application_id, interview_id, payload)
    except ApplicationWorkflowNotFoundError as exc:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(exc))
    except ApplicationWorkflowValidationError as exc:
        raise HTTPException(status_code=status.HTTP_422_UNPROCESSABLE_ENTITY, detail=str(exc))
    except ApplicationWorkflowSeedMissingError as exc:
        raise HTTPException(status_code=status.HTTP_409_CONFLICT, detail=str(exc))


@router.post(
    "/applications/{application_id}/interviews/{interview_id}/cancel",
    response_model=ApplicationInterviewStageResponse,
)
def cancel_application_interview(
    application_id: uuid.UUID,
    interview_id: uuid.UUID,
    payload: ApplicationInterviewCancelRequest | None = None,
    service: ApplicationWorkflowService = Depends(get_application_workflow_service),
):
    try:
        return service.cancel_interview_stage(application_id, interview_id, payload)
    except ApplicationWorkflowNotFoundError as exc:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(exc))
    except ApplicationWorkflowValidationError as exc:
        raise HTTPException(status_code=status.HTTP_422_UNPROCESSABLE_ENTITY, detail=str(exc))
    except ApplicationWorkflowSeedMissingError as exc:
        raise HTTPException(status_code=status.HTTP_409_CONFLICT, detail=str(exc))


@router.delete(
    "/applications/{application_id}/interviews/{interview_id}",
    status_code=status.HTTP_204_NO_CONTENT,
)
def delete_application_interview(
    application_id: uuid.UUID,
    interview_id: uuid.UUID,
    service: ApplicationWorkflowService = Depends(get_application_workflow_service),
):
    try:
        service.delete_interview_stage(application_id, interview_id)
    except ApplicationWorkflowNotFoundError as exc:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(exc))
    except ApplicationWorkflowSeedMissingError as exc:
        raise HTTPException(status_code=status.HTTP_409_CONFLICT, detail=str(exc))


@router.get(
    "/applications/{application_id}/links",
    response_model=list[ApplicationExternalLinkResponse],
)
def list_application_links(
    application_id: uuid.UUID,
    service: ApplicationWorkflowService = Depends(get_application_workflow_service),
):
    try:
        return service.list_external_links(application_id)
    except ApplicationWorkflowNotFoundError as exc:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(exc))
    except ApplicationWorkflowSeedMissingError as exc:
        raise HTTPException(status_code=status.HTTP_409_CONFLICT, detail=str(exc))


@router.post(
    "/applications/{application_id}/links",
    response_model=ApplicationExternalLinkResponse,
    status_code=status.HTTP_201_CREATED,
)
def create_application_link(
    application_id: uuid.UUID,
    payload: ApplicationExternalLinkCreate,
    service: ApplicationWorkflowService = Depends(get_application_workflow_service),
):
    try:
        return service.create_external_link(application_id, payload)
    except ApplicationWorkflowNotFoundError as exc:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(exc))
    except ApplicationWorkflowSeedMissingError as exc:
        raise HTTPException(status_code=status.HTTP_409_CONFLICT, detail=str(exc))


@router.patch(
    "/applications/{application_id}/links/{link_id}",
    response_model=ApplicationExternalLinkResponse,
)
def update_application_link(
    application_id: uuid.UUID,
    link_id: uuid.UUID,
    payload: ApplicationExternalLinkUpdate,
    service: ApplicationWorkflowService = Depends(get_application_workflow_service),
):
    try:
        return service.update_external_link(application_id, link_id, payload)
    except ApplicationWorkflowNotFoundError as exc:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(exc))
    except ApplicationWorkflowSeedMissingError as exc:
        raise HTTPException(status_code=status.HTTP_409_CONFLICT, detail=str(exc))


@router.delete(
    "/applications/{application_id}/links/{link_id}",
    status_code=status.HTTP_204_NO_CONTENT,
)
def delete_application_link(
    application_id: uuid.UUID,
    link_id: uuid.UUID,
    service: ApplicationWorkflowService = Depends(get_application_workflow_service),
):
    try:
        service.delete_external_link(application_id, link_id)
    except ApplicationWorkflowNotFoundError as exc:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(exc))
    except ApplicationWorkflowSeedMissingError as exc:
        raise HTTPException(status_code=status.HTTP_409_CONFLICT, detail=str(exc))


@router.patch("/applications/{application_id}", response_model=ApplicationDetailResponse)
def update_application(
    application_id: uuid.UUID,
    payload: ApplicationMetadataUpdate,
    service: ApplicationWorkflowService = Depends(get_application_workflow_service),
):
    try:
        return service.update_application(application_id, payload)
    except ApplicationWorkflowNotFoundError as exc:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(exc))
    except ApplicationWorkflowValidationError as exc:
        raise HTTPException(status_code=status.HTTP_502_BAD_GATEWAY, detail=str(exc))
    except ApplicationWorkflowSeedMissingError as exc:
        raise HTTPException(status_code=status.HTTP_409_CONFLICT, detail=str(exc))


@router.post(
    "/applications/{application_id}/transition",
    response_model=ApplicationDetailResponse,
)
def transition_application_state_alias(
    application_id: uuid.UUID,
    payload: ApplicationStateTransitionRequest,
    service: ApplicationWorkflowService = Depends(get_application_workflow_service),
):
    return _transition_application_state(application_id, payload, service)


@router.post(
    "/applications/{application_id}/state-transitions",
    response_model=ApplicationDetailResponse,
)
def transition_application_state(
    application_id: uuid.UUID,
    payload: ApplicationStateTransitionRequest,
    service: ApplicationWorkflowService = Depends(get_application_workflow_service),
):
    return _transition_application_state(application_id, payload, service)


def _transition_application_state(
    application_id: uuid.UUID,
    payload: ApplicationStateTransitionRequest,
    service: ApplicationWorkflowService,
):
    try:
        return service.transition_state(application_id, payload)
    except ApplicationWorkflowNotFoundError as exc:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(exc))
    except ApplicationWorkflowTransitionError as exc:
        raise HTTPException(status_code=status.HTTP_409_CONFLICT, detail=str(exc))
    except ApplicationWorkflowValidationError as exc:
        raise HTTPException(status_code=status.HTTP_502_BAD_GATEWAY, detail=str(exc))
    except ApplicationWorkflowSeedMissingError as exc:
        raise HTTPException(status_code=status.HTTP_409_CONFLICT, detail=str(exc))

File: frontend/src/api/applications.ts
import { apiRequest } from "./client";
import type {
  ApplicationDetail,
  ApplicationExternalLink,
  ApplicationExternalLinkPayload,
  ApplicationInterviewCancelPayload,
  ApplicationInterviewCompletePayload,
  ApplicationInterviewStage,
  ApplicationInterviewStagePayload,
  ApplicationNote,
  ApplicationNotePayload,
  ApplicationPipelineResponse,
  ApplicationReminder,
  ApplicationReminderPayload,
  ApplicationStateTransitionPayload,
  ApplicationSummary,
  ApplicationTimelineEvent,
} from "../types/applications";

function appendPipelineParams(
  path: string,
  {
    workspaceId,
    includeInactive,
  }: {
    workspaceId?: string | null;
    includeInactive?: boolean;
  } = {},
) {
  const params = new URLSearchParams();
  if (workspaceId) {
    params.set("workspace_id", workspaceId);
  }
  if (includeInactive) {
    params.set("include_inactive", "true");
  }
  const query = params.toString();
  return query ? `${path}?${query}` : path;
}

export function getApplicationsPipeline(options?: {
  workspaceId?: string | null;
  includeInactive?: boolean;
}): Promise<ApplicationPipelineResponse> {
  return apiRequest<ApplicationPipelineResponse>(
    appendPipelineParams("/api/applications/pipeline", options),
  );
}

export function getWorkspaceApplicationsPipeline(
  workspaceId: string,
  options?: { includeInactive?: boolean },
): Promise<ApplicationPipelineResponse> {
  return apiRequest<ApplicationPipelineResponse>(
    appendPipelineParams(`/api/workspaces/${workspaceId}/applications/pipeline`, {
      includeInactive: options?.includeInactive,
    }),
  );
}

export function getApplication(applicationId: string): Promise<ApplicationDetail> {
  return apiRequest<ApplicationDetail>(`/api/applications/${applicationId}`);
}

export function getApplicationTimeline(
  applicationId: string,
): Promise<ApplicationTimelineEvent[]> {
  return apiRequest<ApplicationTimelineEvent[]>(
    `/api/applications/${applicationId}/timeline`,
  );
}

export function transitionApplicationState(
  applicationId: string,
  payload: ApplicationStateTransitionPayload,
): Promise<ApplicationSummary> {
  return apiRequest<ApplicationSummary>(
    `/api/applications/${applicationId}/state-transitions`,
    {
      method: "POST",
      body: JSON.stringify(payload),
    },
  );
}

export function listApplicationNotes(
  applicationId: string,
): Promise<ApplicationNote[]> {
  return apiRequest<ApplicationNote[]>(`/api/applications/${applicationId}/notes`);
}

export function createApplicationNote(
  applicationId: string,
  payload: ApplicationNotePayload,
): Promise<ApplicationNote> {
  return apiRequest<ApplicationNote>(`/api/applications/${applicationId}/notes`, {
    method: "POST",
    body: JSON.stringify(payload),
  });
}

export function updateApplicationNote(
  applicationId: string,
  noteId: string,
  payload: Partial<ApplicationNotePayload>,
): Promise<ApplicationNote> {
  return apiRequest<ApplicationNote>(
    `/api/applications/${applicationId}/notes/${noteId}`,
    {
      method: "PATCH",
      body: JSON.stringify(payload),
    },
  );
}

export function deleteApplicationNote(
  applicationId: string,
  noteId: string,
): Promise<void> {
  return apiRequest<void>(`/api/applications/${applicationId}/notes/${noteId}`, {
    method: "DELETE",
  });
}

export function listApplicationReminders(
  applicationId: string,
): Promise<ApplicationReminder[]> {
  return apiRequest<ApplicationReminder[]>(
    `/api/applications/${applicationId}/reminders`,
  );
}

export function createApplicationReminder(
  applicationId: string,
  payload: ApplicationReminderPayload,
): Promise<ApplicationReminder> {
  return apiRequest<ApplicationReminder>(
    `/api/applications/${applicationId}/reminders`,
    {
      method: "POST",
      body: JSON.stringify(payload),
    },
  );
}

export function updateApplicationReminder(
  applicationId: string,
  reminderId: string,
  payload: Partial<ApplicationReminderPayload>,
): Promise<ApplicationReminder> {
  return apiRequest<ApplicationReminder>(
    `/api/applications/${applicationId}/reminders/${reminderId}`,
    {
      method: "PATCH",
      body: JSON.stringify(payload),
    },
  );
}

export function completeApplicationReminder(
  applicationId: string,
  reminderId: string,
): Promise<ApplicationReminder> {
  return apiRequest<ApplicationReminder>(
    `/api/applications/${applicationId}/reminders/${reminderId}/complete`,
    {
      method: "POST",
    },
  );
}

export function listApplicationLinks(
  applicationId: string,
): Promise<ApplicationExternalLink[]> {
  return apiRequest<ApplicationExternalLink[]>(
    `/api/applications/${applicationId}/links`,
  );
}

export function createApplicationLink(
  applicationId: string,
  payload: ApplicationExternalLinkPayload,
): Promise<ApplicationExternalLink> {
  return apiRequest<ApplicationExternalLink>(
    `/api/applications/${applicationId}/links`,
    {
      method: "POST",
      body: JSON.stringify(payload),
    },
  );
}

export function updateApplicationLink(
  applicationId: string,
  linkId: string,
  payload: Partial<ApplicationExternalLinkPayload>,
): Promise<ApplicationExternalLink> {
  return apiRequest<ApplicationExternalLink>(
    `/api/applications/${applicationId}/links/${linkId}`,
    {
      method: "PATCH",
      body: JSON.stringify(payload),
    },
  );
}

export function deleteApplicationLink(
  applicationId: string,
  linkId: string,
): Promise<void> {
  return apiRequest<void>(`/api/applications/${applicationId}/links/${linkId}`, {
    method: "DELETE",
  });
}


export function listApplicationInterviews(
  applicationId: string,
): Promise<ApplicationInterviewStage[]> {
  return apiRequest<ApplicationInterviewStage[]>(
    `/api/applications/${applicationId}/interviews`,
  );
}

export function createApplicationInterview(
  applicationId: string,
  payload: ApplicationInterviewStagePayload,
): Promise<ApplicationInterviewStage> {
  return apiRequest<ApplicationInterviewStage>(
    `/api/applications/${applicationId}/interviews`,
    {
      method: "POST",
      body: JSON.stringify(payload),
    },
  );
}

export function updateApplicationInterview(
  applicationId: string,
  interviewId: string,
  payload: Partial<ApplicationInterviewStagePayload>,
): Promise<ApplicationInterviewStage> {
  return apiRequest<ApplicationInterviewStage>(
    `/api/applications/${applicationId}/interviews/${interviewId}`,
    {
      method: "PATCH",
      body: JSON.stringify(payload),
    },
  );
}

export function completeApplicationInterview(
  applicationId: string,
  interviewId: string,
  payload: ApplicationInterviewCompletePayload = {},
): Promise<ApplicationInterviewStage> {
  return apiRequest<ApplicationInterviewStage>(
    `/api/applications/${applicationId}/interviews/${interviewId}/complete`,
    {
      method: "POST",
      body: JSON.stringify(payload),
    },
  );
}

export function cancelApplicationInterview(
  applicationId: string,
  interviewId: string,
  payload: ApplicationInterviewCancelPayload = {},
): Promise<ApplicationInterviewStage> {
  return apiRequest<ApplicationInterviewStage>(
    `/api/applications/${applicationId}/interviews/${interviewId}/cancel`,
    {
      method: "POST",
      body: JSON.stringify(payload),
    },
  );
}

export function deleteApplicationInterview(
  applicationId: string,
  interviewId: string,
): Promise<void> {
  return apiRequest<void>(
    `/api/applications/${applicationId}/interviews/${interviewId}`,
    { method: "DELETE" },
  );
}
